extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,270 @@
+        this.numConcurrentParts = s3Instructions.numConcurrentParts();
+        this.pendingRequests = new ArrayList<>(numConcurrentParts);
+
+        this.nextPartNumber = MIN_PART_NUMBER;
+        this.completedParts = new ArrayList<>();
+    }
+
+    public void write(int b) throws IOException {
+        write(new byte[] {(byte) b}, 0, 1);
+    }
+
+    public void write(byte[] b) throws IOException {
+        write(b, 0, b.length);
+    }
+
+    public void write(final byte @NotNull [] b, int off, int len) throws IOException {
+        while (len != 0) {
+            if (uploadId == null) {
+                // Initialize the upload ID for the multipart upload
+                uploadId = initiateMultipartUpload();
+            }
+
+            // We use buffers and futures in a round-robin fashion
+            final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
+            if (pendingRequests.size() == nextSlotId) {
+                pendingRequests.add(new OutgoingRequest(partSize));
+            } else if (pendingRequests.size() < nextSlotId) {
+                throw new IllegalStateException("Unexpected slot ID " + nextSlotId + " for uri " + uri + " with " +
+                        pendingRequests.size() + " pending requests.");
+            }
+
+            // Wait for the oldest upload to complete if no space is available
+            final OutgoingRequest useRequest = pendingRequests.get(nextSlotId);
+            if (useRequest.future != null) {
+                waitForCompletion(useRequest);
+            }
+
+            // Write as much as possible to this buffer
+            final ByteBuffer buffer = useRequest.buffer;
+            final int remaining = buffer.remaining();
+            if (remaining >= len) {
+                buffer.put(b, off, len);
+                if (!buffer.hasRemaining()) {
+                    sendPartRequest(useRequest);
+                }
+                break; // done
+            }
+            buffer.put(b, off, remaining);
+            sendPartRequest(useRequest);
+            off += remaining;
+            len -= remaining;
+        }
+    }
+
+    public void flush() throws IOException {
+        final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
+        if (pendingRequests.size() == nextSlotId) {
+            // Nothing to flush
+            return;
+        }
+        final OutgoingRequest request = pendingRequests.get(nextSlotId);
+        if (request.buffer.position() != 0 && request.future == null) {
+            sendPartRequest(request);
+        }
+    }
+
+    /**
+     * Try to finish the multipart upload and close the stream. Cancel the upload if an error occurs.
+     *
+     * @throws IOException if an error occurs while closing the stream
+     */
+    public void close() throws IOException {
+        if (uploadId == null) {
+            return;
+        }
+        try {
+            flush();
```

Comment:
Do we think it should ever be the closing logic should ever flush the last request? I would think not?

It may be better to consider an alternative model where the caller explicitly completes the request:

```java
try (final S3OutputStream out = new S3OutputStream(...)) {
  out.write(...);
  out.complete();
}
```

in this way, close has explicit knowledge of whether complete was called or not and can abort only if appropriate.

Reply:
I can do this by adding a new output stream interface which exposes a complete method and parquet code will call complete instead of flush when writing to a file is done.

Reply:
Your proposal is similar to how https://github.com/CI-CMG/aws-s3-outputstream does it.
But there, if we call close on the stream, it completes by default, rather than aborting. So we would need to decide on that default behavior.